### PR TITLE
Implement glue option and fuzzy tolerance in union

### LINKED ISSUE #374: Expose `BRepAlgoAPI_Fuse.SetGlue` in the union operation
For non-overlapping shapes (i.e. gluing two parts), `BRepAlgoAPI_Fuse.SetGlue` results in massive speedups (100x).  Something along those lines
```python
a.union(b,glue=True)
```
Related to #336.

Code example:
```python
import cadquery as cq
import math

Nsegmentsaround, Nsegmentsalong = 50, 15

def cycle(r, rad):
    r = r.moveTo(0, 0)
    for i in range(1, Nsegmentsaround):
        th = math.radians(i/Nsegmentsaround*360)
        x, y = rad*(1-math.cos(th)), 0.2*math.sin(th)*rad
        r = r.lineTo(x, y)
    r = r.close()
    return r

r = cq.Workplane("XZ", (0,0,0))
r = cycle(r, 0.6)
for j in range(Nsegmentsalong):
    r = r.workplane(offset=0.2, centerOption="ProjectedOrigin")
    r = cycle(r, 1/(j+2))
r = r.loft(combine=True)

m = r.mirror(mirrorPlane='XZ', basePointVector=(0,0,0))
res = m.union(r)  # <-- this takes too much time as Nsegmentsalong increases

import OCP

args = OCP.TopTools.TopTools_ListOfShape()
args.Append(m.val().wrapped)
tools = OCP.TopTools.TopTools_ListOfShape()
tools.Append(r.val().wrapped)

fop = OCP.BRepAlgoAPI.BRepAlgoAPI_Fuse()
fop.SetArguments(args)
fop.SetTools(tools)

fop.SetGlue(OCP.BOPAlgo.BOPAlgo_GlueEnum.BOPAlgo_GlueShift)
```

Result with:
```
%time fop.Build()
CPU times: user 258 ms, sys: 17 µs, total: 258 ms
Wall time: 256 ms
```
Result without
```
%time fop.Build()
CPU times: user 27.6 s, sys: 106 ms, total: 27.7 s
Wall time: 26.7 s
```

### PR DESCRIPTION
**Pull request by adam-urbanczyk**
Will resolve #374 . It will also enable advanced users to apply the fuzzy bool ops (see #184 ).

### CODE REVIEW

**Comment by bragostin** [comment]
For example consider the following example taken from Ex101 where interpPlates are unioned together:
```python
# EXAMPLE 3
# Points on hexagonal pattern coordinates, use of pushpoints.
r1 = 1.0
N = 3
ca = cos(30.0 * pi / 180.0)
sa = sin(30.0 * pi / 180.0)
# EVEN ROWS
pts = [
    (-3.0, -3.0),
    (-1.267949, -3.0),
    (0.464102, -3.0),
    (2.196152, -3.0),
    (-3.0, 0.0),
    (-1.267949, 0.0),
    (0.464102, 0.0),
    (2.196152, 0.0),
    (-2.133974, -1.5),
    (-0.401923, -1.5),
    (1.330127, -1.5),
    (3.062178, -1.5),
    (-2.133975, 1.5),
    (-0.401924, 1.5),
    (1.330127, 1.5),
    (3.062178, 1.5),
]

# Spike surface
thickness = 0.1
fn = 6
edge_points = [
    [
        r1 * cos(i * 2 * pi / fn + 30 * pi / 180),
        r1 * sin(i * 2 * pi / fn + 30 * pi / 180),
    ]
    for i in range(fn + 1)
]
surface_points = [
    [
        r1 / 4 * cos(i * 2 * pi / fn + 30 * pi / 180),
        r1 / 4 * sin(i * 2 * pi / fn + 30 * pi / 180),
        0.75,
    ]
    for i in range(fn + 1)
] + [[0, 0, 2]]
edge_wire = cq.Workplane("XY").polyline(edge_points)
plate_3 = (
    cq.Workplane("XY")
    .pushPoints(pts)
    .interpPlate(
        edge_wire,
        surface_points,
        thickness,
        combine=False,
        clean=False,
        degree=2,
        nbPtsOnCur=20,
        nbIter=2,
        anisotropy=False,
        tol2d=0.00001,
        tol3d=0.0001,
        tolAng=0.01,
        tolCurv=0.1,
        maxDeg=8,
        maxSegments=9,
    )
)

# TODO

plate_3 = [o.vals()[0] for o in plate_3.all()]
print("plate_3 = ", plate_3)
plate_3 = cq.Compound.makeCompound(plate_3)
print("plate_3 = ", plate_3)
#plate_3.exportStep('plate_3.stp')
plate_3 = cq.Workplane("XY").union(cq.CQ(plate_3), glue=True, tol=1e-2)
plate_3.val().exportStep('plate_3.stp')

# TODO
print("plate_3.val().Volume() = ", plate_3.val().Volume())
plate_3 = plate_3.translate((0, 4 * 11, 0))
show_object(plate_3)
```
It crashes with standard ``union`` and also gluing without tolerance. However it works almost instantly with gluing and the Fuzzy tolerance set at 0.01.

**Comment by adam-urbanczyk** [comment]
I added the tol option @bragostin . Nice model BTW:
obraz

File: cadquery/cq.py [hunk 1]
```diff
@@ -2715,14 +2715,15 @@ def combine(self, clean=True):
...
+    def union(self, toUnion=None, clean=True, glue=False, tol=None):
         """
         Unions all of the items on the stack of toUnion with the current solid.
         If there is no current solid, the items in toUnion are unioned together.
         
         :param toUnion:
         :type toUnion: a solid object, or a CQ object having a solid,
         :param boolean clean: call :py:meth:`clean` afterwards to have a clean shape
+        :param boolean glue: use a faster gluing mode for non-overlapping shapes
         :raises: ValueError if there is no solid to add to in the chain
```
Comment (jmwright) [inline cadquery/cq.py:2727]: Seems to be missing `tol` parameter in doc string. `tol` being `None` won't break anything? Does it just let OCCT auto-select the tolerance in that case?

File: cadquery/occ_impl/shapes.py [hunk 2]
```diff
@@ -608,13 +610,16 @@ def cut(self, *toCut):
...
+    def fuse(self, *toFuse, glue=False, tol=None):
         """
         Fuse shapes together
         """
 
         fuse_op = BRepAlgoAPI_Fuse()
-        # fuse_op.SetFuzzyValue(TOLERANCE)
+        if glue:
+            fuse_op.SetGlue(BOPAlgo_GlueEnum.BOPAlgo_GlueShift)
+        if tol:
```
Comment (jmwright) [inline cadquery/occ_impl/shapes.py:621]: Ok, I see how you're using `tol` now.

File: cadquery/cq.py (hunk 1 above)
Comment (adam-urbanczyk) [reply cadquery/cq.py:2727]: Good catch. Adding the docstring.

**Comment by adam-urbanczyk** [comment]
Docstring is fixed, if CI passes the PR is ready for merging.

**Comment by jmwright** [comment]
@adam-urbanczyk I went ahead and merged since everything looked fine.

**Pull request merged**